pyquil/api/_quantum_computer.py

```diff
@@ -411,17 +417,34 @@ def run_and_measure(self, program: Program, trials: int) -> Dict[int, np.ndarray
         :return: A dictionary keyed by qubit index where the corresponding value is a 1D array of
             measured bits.
         """
+
+        def unwrap_qubit(q: QubitDesignator) -> int:
+            if isinstance(q, Qubit):
+                return q.index
+            elif isinstance(q, int):
+                return q
+            else:
+                raise TypeError(
+                    "Cannot run and measure program with " f"unaddressed QubitPlaceholders: {q}"
+                )
+
         program = program.copy()
         validate_supported_quil(program)
         ro = program.declare("ro", "BIT", len(self.qubits()))
-        for i, q in enumerate(self.qubits()):
+        measure_all = isinstance(self.qam, QVM) and self.qam.noise_model is not None
```

Comment:
This appears to me to disagree with the docstring:

Docstring: "If a noiseless QVM, then measure subset; otherwise, measure all"

Here: "If a noisy QVM, measure all; otherwise measure subset"

The difference: a non-QVM. Per docstring, will measure all. Per this, will measure only the program's qubits. That seems like an unintended change?

maybe `measure_all = not (isinstance(self.qam, QVM) and self.qam.noise_model is not None)` ?